fix (jkube-kit/resource/helm) : Remove classifier from Helm Chart tarball name (#1369)

rohanKanojia opened the pull request:
## Description
Fixes #1369

+ Add new field in HelmConfig with name tarFileClassifier (defaults to empty string)
+ Change default location of Helm Chart tarball directory to output
  directory to avoid collission between K8s and OpenShift helm chart
  tarballs.

### Breaking Changes:
- Helm archive generated by `k8s:helm`/`k8sHelm` would no longer contain trailing classifier by default. However, you can add a classifier using `jkube.helm.tarFileClassifier` configuration field. For example, In previous releases a helm chart named `foo` with version `0.0.1` had `foo-0.0.1-helm.tar.gz` as archive name. From this release onwards, the name for equivalent archive file would be `foo-0.0.1.tar.gz`
- Default location of Helm archive generated by `k8s:helm`/`k8sHelm` has been changed. In previous releases, it was (`target`/`build`). From this release onwards, Helm Chart archive would be generated in the same directory where Helm Chart is generated, defaulting to the following values:
  - Kubernetes Maven Plugin : `target/jkube/helm/${chartName}/kubernetes`
  - Kubernetes Gradle Plugin : `build/jkube/helm/${chartName}/kubernetes`
  - OpenShift Maven Plugin : `target/jkube/helm/${chartName}/openshift`
  - OpenShift Gradle Plugin : `build/jkube/helm/${chartName}/openshift`

Signed-off-by: Rohan Kumar <rohaan@redhat.com>



## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [X] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X] I have performed a self-review of my code
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/jkubeio/jkube-integration-tests)
Please check integration tests and provide/improve tests if necessary.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your issue as ready
-->

manusa commented:
<!-- Eclipse JKube CI bot report -->

rohanKanojia commented:
E2E tests need to be updated

manusa commented:
Relates to https://github.com/dekorateio/dekorate/pull/1087

rohanKanojia commented:
@manusa : https://github.com/dekorateio/dekorate/pull/1087 seems to make classifier a configurable field with default value as EMPTY string.

Shall I modify this PR to add a configurable field for the classifier as well?

manusa commented:
> @manusa : [dekorateio/dekorate#1087](https://github.com/dekorateio/dekorate/pull/1087) seems to make classifier a configurable field with default value as EMPTY string.
> 
> Shall I modify this PR to add a configurable field for the classifier as well?

Yes, I think that would keep things consistent in case of an eventual merge + also allow for backwards compatibility in our side.

rohanKanojia commented:
I checked using a Quarkus project. There is some difference in locations where jkube and dekorate generate helm charts:

**JKube:**
Helm chart location: target/jkube/helm/projectName/kubernetes/
Helm chart archive location: target

**Dekorate:**
Helm chart location: target/helm/kubernetes/projectName
Helm chart archive location: target/helm/kubernetes (disable by default)

Do we want to jkube to change it's default locations to be same as dekorate?

manusa commented on jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java:
```diff
@@ -323,4 +324,11 @@ private static List<HelmParameter> collectParameters(HelmConfig helmConfig) {
   }
+
+  private static String resolveHelmClassifier(HelmConfig helmConfig) {
+    if (StringUtils.isEmpty(helmConfig.getTarFileClassifier())) {
```
note for any subsequent PR, please use isBlank instead, this way we make sure a classifier such as `  ` is not considered.

manusa commented on jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmServiceIT.java:
```diff
@@ -47,12 +47,16 @@ class HelmServiceIT {
   private HelmService helmService;
   private HelmConfig helmConfig;
   private File helmOutputDir;
+  private static File helmK8sOutputDir;
+  private static File helmOpenShiftOutputDir;
```
why are these variables static??

rohanKanojia commented:
@manusa : Helm chart was getting generated in `helm-output` directory because we had tarballOutputDirectory hardcoded in `setUp` method:

https://github.com/eclipse/jkube/blob/33a72d2d99817924a3a67c5caf12df8d176c7f75/jkube-kit/resource/helm/src/test/java/org/eclipse/jkube/kit/resource/helm/HelmServiceIT.java#L59

Previously, In order to change the default `tarballOutputDir`, I had just changed default value here:
https://github.com/eclipse/jkube/blob/33a72d2d99817924a3a67c5caf12df8d176c7f75/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmServiceUtil.java#L102

This was causing problems with the case when there was more than one helm type(HelmType.KUBERNETES and HelmType.OPENSHIFT). I have moved tarballOutDir defaulting logic to HelmService so that we can use the helmType provided in that particular helm chart generation iteration.
